polar-core/src/normalize.rs

```diff
@@ -0,0 +1,317 @@
+use crate::terms::*;
+
+impl Term {
+    /// intelligently negate an expression
+    fn negated(&self) -> Self {
+        use {Operator::*, Value::*};
+        match self.value() {
+            // negate booleans
+            Boolean(b) => self.clone_with_value(Boolean(!b)),
+
+            Expression(Operation { operator, args }) => match operator {
+                // cancel double negatives
+                Not => self.clone_with_value(args[0].value().clone()),
+
+                // swap and/or using De Morgan's to keep the nots nested
+                // as deep as possible
+                And => {
+                    assert_eq!(args.len(), 2);
```

Comment:
Is this guaranteed to be true? At least in the parser side of things we kind of "flatten" arguments to `or` and `and` so that they may have N args each, wondering if that might ever be the case here

Reply:
`btnf` takes care of that 😃